daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala

```diff
@@ -592,18 +620,39 @@ private[lf] case class PartialTransaction(
   def isConsumed(coid: Value.ContractId): Boolean = consumedBy.contains(coid)
 
   /** Guard the execution of a step with the unconsumedness of a
-    * `ContractId`
+    * `ContractId` and check that there is no other contract with the
+    *  same contract key.
```

Comment:
This feels like a strange place to check for key collisions.

The only statement that can introduce a key collision is a create, which is where I'd expect the check to be.

The updates of the `keys` map should happen in the respective mutating commands: create (insert key), archive (set key to None), fetch (insert key if present), lookupByKey (insert result in key map).

Reply:
> The only statement that can introduce a key collision is a create, which is where I'd expect the check to be.

This is not true. I can have a `create` first and then a `fetch` of a global contract with the same key. Yes the ledger should also reject this but this is still an internal duplicate key error where there are two contracts with the same key active at the same time and imho speedy should reject this.

Reply:
Here our mental models disagree. What I'd expect speedy to do is to read the key from the ledger as part of the create so that it can check that the key is not already assigned.

Reply:
I’m open to rejecting this case, see my comment below. I don’t think we should reject a single `create` without the `fetch` later. This is an internally consistent transaction as in there exists a ledger state it could be applied to. The ledger should make sure that the transaction being created by speedy can actually be applied to the current ledger state.

Reply:
Thinking more about my proposal, I believe:
1. it works to detect contract key inconsistencies between the transaction and any contracts whose stakeholders intersect with `actAs \union readAs`
2. it does not detect contract key inconsistencies between the transaction and divulged contracts visible to `actAs \union readAs` that are fetched within the transaction

I'm wondering whether we should aim to detect the conflicts from Case 2, as divulgence is going to go away #9283.

@cocreature @andreaslochbihler-da what do you think?

Reply:
If I understand you correctly, you are suggesting that instead of doing the validation on every fetch & exercise you only do it on a global fetch or exercise? I don’t understand how that affects divulgence. A global fetch could still fetch a divulged contract.

I actually started out with that approach and abandoned for a few reasons:
1. I generally would like to limit different treatment of global & local operations as much as possible. This is exactly how we end up with these type of errors in the first place. I trust the solution here much more to catch all cases and to be robust against future changes.
2. In addition to not making it safer, it doesn’t seem to make anything simpler, you still have to check fetches and exercises which we already check. But now you need logic to track local contracts in `PartialTransaction` and you have some extra code in fetches/exercises to skip the check (which stays exactly the same we have atm).

Reply:
Yes, having made another pass on your implementation I agree. Thanks for your kind explanations.